pkg/operator/controller/dns_status.go

```diff
@@ -92,34 +170,109 @@ func computeDNSDegradedCondition(oldCondition *operatorv1.OperatorCondition, clu
 		maxUnavailable, intstrErr := intstr.GetScaledValueFromIntOrPercent(&maxUnavailableIntStr, int(want), true)
 		switch {
 		case want == 0:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "NoDNSPodsDesired")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSNoDNSPodsDesired, Status: operatorv1.ConditionTrue, LastTransitionTime: now})
 			messages = append(messages, "No DNS pods are desired; this could mean all nodes are tainted or unschedulable.")
 		case have == 0:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "NoDNSPodsAvailable")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSNoDNSPodsAvailable, Status: operatorv1.ConditionTrue, LastTransitionTime: now})
 			messages = append(messages, "No DNS pods are available.")
 		case intstrErr != nil:
-			degradedReasons = append(degradedReasons, "InvalidDNSMaxUnavailable")
+			keepConditionUnknown = true
```

Comment:
This logic has a subtle problem.  With the old logic, when we reach this case, `status` could be "True" if the `if len(clusterIP) == 0 {` check above were true, and otherwise `status` would have its initial value of `Unknown`; either way, this `case intstrErr != nil:` check would leave `status` set to the one of those two values (true or unknown) that it had before this check.  With the new logic, this `case intstrErr != nil:` check always sets `keepConditionUnknown = true` and thus causes the `if keepConditionUnknown {` check below to set `Status` to "Unknown", even if the `if len(clusterIP) == 0 {` check above was true.  

The idea in the original logic is that starting out, we don't know the status, so we set it to "unknown"; then we check for various possible problems; each check can determine that something is wrong and set the degraded status to "true"; and if a previous check has determined that something is wrong, a subsequent check cannot not set the state _back_ to "unknown" because we _do_ know (from the previous check) that _something_ is wrong.  Once we know something is wrong, there is no circumstance in which we can go back to saying it's unknown whether anything is wrong.

Reply:
The fact that this subtle problem wasn't caught by the unit tests suggests we need a test case for "no clusterIP, DNS invalid MaxUnavailable"; such a test case should expect `Degraded=True`, and if I'm reading the new logic correctly, it would fail.

Reply:
Done - fixed other test cases too, which used to look for ConditionUnknown